components/patient-data/impl/src/main/java/org/phenotips/data/internal/controller/GeneListController.java

```diff
@@ -183,4 +189,78 @@ public void writeJSON(Patient patient, JSONObject json, Collection<String> selec
             }
         }
     }
+
+    @Override
+    public PatientData<Map<String, String>> readJSON(JSONObject json)
+    {
+        Object testGeneJson = json.get(this.getJsonPropertyName());
+        if (testGeneJson == null) {
+            return null;
+        }
+
+        try {
+            JSONArray genesJson = json.getJSONArray(this.getJsonPropertyName());
+            List<Map<String, String>> allGenes = new LinkedList<Map<String, String>>();
+            for (Object geneJsonUncast : genesJson) {
+                JSONObject geneJson = JSONObject.fromObject(geneJsonUncast);
+                Map<String, String> singleGene = new LinkedHashMap<String, String>();
+                for (String property : getProperties()) {
+                    // todo. it makes no sense to have no gene name, but to have comments, which can occur
+                    if (geneJson.has(property)) {
+                        String field = geneJson.getString(property);
+                        if (field != null) {
+                            singleGene.put(property, field);
+                        }
+                    }
+                }
+                allGenes.add(singleGene);
+            }
+
+            if (allGenes.isEmpty()) {
+                return null;
+            } else {
+                return new IndexedPatientData<Map<String, String>>(getName(), allGenes);
+            }
+        } catch (Exception e) {
+            this.logger.error("Could not load genes from JSON", e.getMessage());
+        }
+        return null;
+    }
+
+    @Override
+    public void save(Patient patient)
+    {
+        try {
+            XWikiDocument doc = (XWikiDocument) this.documentAccessBridge.getDocument(patient.getDocument());
+            if (doc == null) {
+                throw new NullPointerException(ERROR_MESSAGE_NO_PATIENT_CLASS);
+            }
+
+            // todo. does not check for existing genes
+            PatientData<Map<String, String>> genes = patient.getData(getName());
```

Comment:
We use `this.` only for field access, not for methods. It's clear that a method without an explicit object to call on belongs to the current object, but a variable can either be a class field, a method argument, or a local variable. Explicitly prefixing fields makes it clear that we're accessing the object state. It can also help avoid bugs if we have two things with the same name, a field and an argument.

Anyway, I never write it manually, I set up Eclipse to do it for me: Window -> Preferences -> Java -> Editor -> Save Actions -> Add 'this' qualifier to unqualified field accesses

![save-actions](https://cloud.githubusercontent.com/assets/88663/11610530/931e62fa-9b73-11e5-96a2-420baf9a95d3.png)